substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -304,6 +304,27 @@ pub mod pallet {
 			Ok(())
 		}
 
+		/// Remove an account from being an `Agent`.
+		///
+		/// This can only be called if the agent has no delegated funds, no pending slashes and no
+		/// unclaimed withdrawals.
+		pub fn remove_agent(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let ledger = AgentLedger::<T>::get(&who).ok_or(Error::<T>::NotAgent)?;
+
+			ensure!(
+				ledger.total_delegated == Zero::zero() &&
+					ledger.pending_slash == Zero::zero() &&
+					ledger.unclaimed_withdrawals == Zero::zero(),
+				Error::<T>::NotAllowed
+			);
+
+			// remove provider reference
+			let _ = frame_system::Pallet::<T>::dec_providers(&who).defensive();
```

Comment:
I'm wondering if this is the best way to deal with a potential error returned from `dec_providers`. Maybe we should return earlier and not remove the agent entry from the map unless the providers were decreased?

OTOH, if there's an error, the agent will linger and the pool may not be destroyed, so probably the best way is to keep this defensive and then fix the inconsistency with a simple migration or something along those lines. wdyt?